src/sentry/models/release.py

```diff
@@ -213,6 +213,34 @@ def merge(cls, to_release, from_releases):
 
             release.delete()
 
+    @classmethod
+    def get_closest_releases(cls, project, start_version, limit=5):
+        # given a release version + project, return next
+        # `limit` releases (includes the release specified by `version`)
+        try:
+            release_dates = cls.objects.filter(
+                organization_id=project.organization_id,
+                version=start_version,
+                projects=project,
+            ).values('date_released', 'date_added').get()
+        except cls.DoesNotExist:
+            return []
+
+        start_date = release_dates['date_released'] or release_dates['date_added']
+
+        return list(Release.objects.filter(
+            projects=project,
+            organization_id=project.organization_id,
+        ).extra(select={
+                'date': 'COALESCE(date_released, date_added)',
+                }
+                ).extra(
+            where=["COALESCE(date_released, date_added) >= %s"],
```

Comment:
This concerns me for database performance. Since this doesn't have an index on this.

Can you get me the raw SQL for this QuerySet and ideally the EXPLAIN? And we can check it out. I feel like this will be a scan over all releases the org has always. And this may be a lot of data for some orgs. Especially since we're not really bounded by time here.

Reply:
the raw SQL is:
```sql
SELECT (COALESCE(date_released, date_added)) AS "date",
       "sentry_release"."id",
       "sentry_release"."organization_id",
       "sentry_release"."project_id",
       "sentry_release"."version",
       "sentry_release"."ref",
       "sentry_release"."url",
       "sentry_release"."date_added",
       "sentry_release"."date_started",
       "sentry_release"."date_released",
       "sentry_release"."data",
       "sentry_release"."new_groups",
       "sentry_release"."owner_id",
       "sentry_release"."commit_count",
       "sentry_release"."last_commit_id",
       "sentry_release"."authors",
       "sentry_release"."total_deploys",
       "sentry_release"."last_deploy_id"
FROM "sentry_release"
INNER JOIN "sentry_release_project" ON ("sentry_release"."id" = "sentry_release_project"."release_id")
WHERE ("sentry_release"."organization_id" = 1
       AND "sentry_release_project"."project_id" = 7
       AND (COALESCE(date_released, date_added) >= '2017-09-01 21:11:10.000414'))
ORDER BY "date" ASC LIMIT 5
```
and the results of the explain are:

```
                                                                         QUERY PLAN                                                                         
------------------------------------------------------------------------------------------------------------------------------------------------------------
 Limit  (cost=49.77..49.78 rows=5 width=746)
   ->  Sort  (cost=49.77..49.82 rows=22 width=746)
         Sort Key: (COALESCE(sentry_release.date_released, sentry_release.date_added))
         ->  Hash Join  (cost=15.44..49.40 rows=22 width=746)
               Hash Cond: (sentry_release.id = sentry_release_project.release_id)
               ->  Seq Scan on sentry_release  (cost=0.00..30.88 rows=286 width=746)
                     Filter: ((COALESCE(date_released, date_added) >= '2017-09-01 21:11:10.000414-07'::timestamp with time zone) AND (organization_id = 1))
               ->  Hash  (cost=14.61..14.61 rows=66 width=8)
                     ->  Bitmap Heap Scan on sentry_release_project  (cost=4.79..14.61 rows=66 width=8)
                           Recheck Cond: (project_id = 7)
                           ->  Bitmap Index Scan on sentry_release_project_project_id  (cost=0.00..4.77 rows=66 width=0)
                                 Index Cond: (project_id = 7)
```

Reply:
Yeah, so this is relative expensive query. What's happening is all rows for a release across the entire org are being pulled into memory, then sorted. In some cases this might be high thousands of rows. It'd depend on how many releases there are. If an org has hundreds of projects each doing a few deploys per day.. this can add up pretty quickly.

So really, the LIMIT here doesn't help any. It just reduces how many rows are returned from the database, but the database still does all the work of sorting.

Is there anything else we can do here? Do we need to COALESCE between those two fields? Could we maybe add an index to one and backfill/remove the other? Or we can see what this looks like with a functional index.

Reply:
for reference, talked to matt and added this index: https://github.com/getsentry/sentry/pull/6090